libvast/include/vast/pipeline.hpp

```diff
@@ -0,0 +1,274 @@
+    // Self&`, and `const operator_control_plane&` instead of
+    // `operator_control_plane&` to detect erroneous definitions.
+    auto f = detail::overload{
+      [&](std::monostate) -> caf::expected<operator_output> {
+        constexpr auto source = std::is_invocable_v<Self&>;
+        constexpr auto source_ctrl
+          = std::is_invocable_v<Self&, const operator_control_plane&>;
+        static_assert(source + source_ctrl <= 1,
+                      "ambiguous operator definition: callable with both "
+                      "`op()` and `op(ctrl)`");
+        if constexpr (source) {
+          return self()();
+        } else if constexpr (source_ctrl) {
+          return self()(ctrl);
+        } else {
+          return caf::make_error(ec::type_clash,
+                                 fmt::format("'{}' cannot be used as a source",
+                                             to_string()));
+        }
+      },
+      [&]<class Input>(
+        generator<Input> input) -> caf::expected<operator_output> {
+        constexpr auto one = std::is_invocable_v<Self&, Input>;
+        constexpr auto gen = std::is_invocable_v<Self&, generator<Input>>;
+        constexpr auto gen_ctrl
+          = std::is_invocable_v<Self&, generator<Input>,
+                                const operator_control_plane&>;
+        static_assert(one + gen + gen_ctrl <= 1,
+                      "ambiguous operator definition: callable with more than "
+                      "one of `op(x)`, `op(gen)` and `op(gen, ctrl)`");
+        if constexpr (one) {
+          return std::invoke(
+            [this](generator<Input> input)
+              -> generator<std::invoke_result_t<Self&, Input>> {
+              for (auto&& x : input) {
+                co_yield self()(std::move(x));
+              }
+            },
+            std::move(input));
+        } else if constexpr (gen) {
+          return self()(std::move(input));
+        } else if constexpr (gen_ctrl) {
+          return self()(std::move(input), ctrl);
+        } else {
+          return caf::make_error(ec::type_clash,
+                                 fmt::format("'{}' does not accept {} as input",
+                                             to_string(),
+                                             operator_element_name<Input>()));
+        }
+      },
+    };
+    return std::visit(f, std::move(input));
+  }
+
+  auto copy() const -> operator_ptr final {
+    return std::make_unique<Self>(self());
+  }
+
+private:
+  auto self() const -> const Self& {
+    static_assert(std::is_final_v<Self>);
+    static_assert(std::is_base_of_v<crtp_operator, Self>);
+    return static_cast<const Self&>(*this);
+  }
+};
+
+/// Pipeline operator with a per-schema initialization.
+///
+/// Usage: Override `initialize` and `process`, perhaps `finish`.
+template <class Self, class State, class Output = table_slice>
+class schematic_operator : public crtp_operator<Self> {
```

Comment:
Good point. However, I prefer to keep this as-is, because it is also needed for `copy` (and perhaps some virtual `inspect`, though I'm not sure whether that's viable).